File: code/client/munkilib/admin/makecatalogslib.py
# encoding: utf-8
#
# Copyright 2017-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
makecatalogslib

Created by Greg Neagle on 2017-11-19.
Routines used by makecatalogs
"""
from __future__ import absolute_import, print_function

# std libs
import hashlib
import os

# our libs
from .common import list_items_of_kind, AttributeDict

from .. import munkirepo

from ..wrappers import readPlistFromString, writePlistToString


class MakeCatalogsError(Exception):
    '''Error to raise when there is problem making catalogs'''
    pass


def hash_icons(repo, output_fn=None):
    '''Builds a dictionary containing hashes for all our repo icons'''
    errors = []
    icons = {}
    if output_fn:
        output_fn("Getting list of icons...")
    icon_list = repo.itemlist('icons')
    # Don't hash the hashes, they aren't icons.
    if '_icon_hashes.plist' in icon_list:
        icon_list.remove('_icon_hashes.plist')
    for icon_ref in icon_list:
        if output_fn:
            output_fn("Hashing %s..." % (icon_ref))
        # Try to read the icon file
        try:
            icondata = repo.get('icons/' + icon_ref)
            icons[icon_ref] = hashlib.sha256(icondata).hexdigest()
        except munkirepo.RepoError as err:
            errors.append(u'RepoError for %s: %s' % (icon_ref, err))
        except IOError as err:
            errors.append(u'IO error for %s: %s' % (icon_ref, err))
        except BaseException as err:
            errors.append(u'Unexpected error for %s: %s' % (icon_ref, err))
    return icons, errors


def verify_pkginfo(pkginfo_ref, pkginfo, pkgs_list, errors):
    '''Returns True if referenced installer items are present,
    False otherwise. Adds errors/warnings to the errors list'''
    installer_type = pkginfo.get('installer_type')
    if installer_type in ['nopkg', 'apple_update_metadata']:
        # no associated installer item (pkg) for these types
        return True
    if pkginfo.get('PackageCompleteURL') or pkginfo.get('PackageURL'):
        # installer item may be on a different server
        return True

    if not 'installer_item_location' in pkginfo:
        errors.append("WARNING: %s is missing installer_item_location"
                      % pkginfo_ref)
        return False

    # Try to form a path and fail if the
    # installer_item_location is not a valid type
    try:
        installeritempath = os.path.join(
            "pkgs", pkginfo['installer_item_location'])
    except TypeError:
        errors.append("WARNING: invalid installer_item_location in %s"
                      % pkginfo_ref)
        return False

    # Check if the installer item actually exists
    if not installeritempath in pkgs_list:
        # do a case-insensitive comparison
        found_caseinsensitive_match = False
        for repo_pkg in pkgs_list:
            if installeritempath.lower() == repo_pkg.lower():
                errors.append(
                    "WARNING: %s refers to installer item: %s. "
                    "The pathname of the item in the repo has "
                    "different case: %s. This may cause issues "
                    "depending on the case-sensitivity of the "
                    "underlying filesystem."
                    % (pkginfo_ref,
                       pkginfo['installer_item_location'], repo_pkg))
                found_caseinsensitive_match = True
                break
        if not found_caseinsensitive_match:
            errors.append(
                "WARNING: %s refers to missing installer item: %s"
                % (pkginfo_ref, pkginfo['installer_item_location']))
            return False

    #uninstaller sanity checking
    uninstaller_type = pkginfo.get('uninstall_method')
    if uninstaller_type in ['AdobeCCPUninstaller']:
        # uninstaller_item_location is required
        if not 'uninstaller_item_location' in pkginfo:
            errors.append(
                "WARNING: %s is missing uninstaller_item_location"
                % pkginfo_ref)
            return False

    # if an uninstaller_item_location is specified, sanity-check it
    if 'uninstaller_item_location' in pkginfo:
        try:
            uninstalleritempath = os.path.join(
                "pkgs", pkginfo['uninstaller_item_location'])
        except TypeError:
            errors.append("WARNING: invalid uninstaller_item_location "
                          "in %s" % pkginfo_ref)
            return False

        # Check if the uninstaller item actually exists
        if not uninstalleritempath in pkgs_list:
            # do a case-insensitive comparison
            found_caseinsensitive_match = False
            for repo_pkg in pkgs_list:
                if uninstalleritempath.lower() == repo_pkg.lower():
                    errors.append(
                        "WARNING: %s refers to uninstaller item: %s. "
                        "The pathname of the item in the repo has "
                        "different case: %s. This may cause issues "
                        "depending on the case-sensitivity of the "
                        "underlying filesystem."
                        % (pkginfo_ref,
                           pkginfo['uninstaller_item_location'], repo_pkg))
                    found_caseinsensitive_match = True
                    break
            if not found_caseinsensitive_match:
                errors.append(
                    "WARNING: %s refers to missing uninstaller item: %s"
                    % (pkginfo_ref, pkginfo['uninstaller_item_location']))
                return False

    # if we get here we passed all the checks
    return True


def process_pkgsinfo(repo, options, output_fn=None):
    '''Processes pkginfo files and returns a dictionary of catalogs'''
    errors = []
    catalogs = {}
    # get a list of pkgsinfo items
    if output_fn:
        output_fn("Getting list of pkgsinfo...")
    try:
        pkgsinfo_list = list_items_of_kind(repo, 'pkgsinfo')
    except munkirepo.RepoError as err:
        raise MakeCatalogsError(
            u"Error getting list of pkgsinfo items: %s" % err)

    # get a list of pkgs items
    if output_fn:
        output_fn("Getting list of pkgs...")
    try:
        pkgs_list = list_items_of_kind(repo, 'pkgs')
    except munkirepo.RepoError as err:
        raise MakeCatalogsError(
            u"Error getting list of pkgs items: %s" % err)

    # start with empty catalogs dict
    catalogs = {}
    catalogs['all'] = []

    # Walk through the pkginfo files
    for pkginfo_ref in pkgsinfo_list:
        # Try to read the pkginfo file
        try:
            data = repo.get(pkginfo_ref)
            pkginfo = readPlistFromString(data)
        except IOError as err:
            errors.append("IO error for %s: %s" % (pkginfo_ref, err))
            continue
        except BaseException as err:
            errors.append("Unexpected error for %s: %s" % (pkginfo_ref, err))
            continue

        if not 'name' in pkginfo:
            errors.append("WARNING: %s is missing name" % pkginfo_ref)
            continue

        # don't copy admin notes to catalogs.
        if pkginfo.get('notes'):
            del pkginfo['notes']
        # strip out any keys that start with "_"
        # (example: pkginfo _metadata)
        for key in list(pkginfo.keys()):
            if key.startswith('_'):
                del pkginfo[key]

        # sanity checking
        if not options.skip_payload_check:
            verified = verify_pkginfo(pkginfo_ref, pkginfo, pkgs_list, errors)
            if not verified and not options.force:
                # Skip this pkginfo unless we're running with force flag
                continue

        # append the pkginfo to the relevant catalogs
        catalogs['all'].append(pkginfo)
        for catalogname in pkginfo.get("catalogs", []):
            if not catalogname:
                errors.append("WARNING: %s has an empty catalogs array!"
                              % pkginfo_ref)
                continue
            if not catalogname in catalogs:
                catalogs[catalogname] = []
            catalogs[catalogname].append(pkginfo)
            if output_fn:
                output_fn("Adding %s to %s..." % (pkginfo_ref, catalogname))

    # look for catalog names that differ only in case
    duplicate_catalogs = []
    for key in catalogs:
        if key.lower() in [item.lower() for item in catalogs if item != key]:
            duplicate_catalogs.append(key)
    if duplicate_catalogs:
        errors.append("WARNING: There are catalogs with names that differ only "
                      "by case. This may cause issues depending on the case-"
                      "sensitivity of the underlying filesystem: %s"
                      % duplicate_catalogs)

    return catalogs, errors


def makecatalogs(repo, options, output_fn=None):
    '''Assembles all pkginfo files into catalogs.
    User calling this needs to be able to write to the repo/catalogs
    directory.'''

    if isinstance(options, dict):
        options = AttributeDict(options)

    icons, errors = hash_icons(repo, output_fn=output_fn)

    catalogs, catalog_errors = process_pkgsinfo(
        repo, options, output_fn=output_fn)

    errors.extend(catalog_errors)

    # clear out old catalogs
    try:
        catalog_list = repo.itemlist('catalogs')
    except munkirepo.RepoError:
        catalog_list = []
    for catalog_name in catalog_list:
        if catalog_name not in list(catalogs.keys()):
            catalog_ref = os.path.join('catalogs', catalog_name)
            try:
                repo.delete(catalog_ref)
            except munkirepo.RepoError:
                errors.append('Could not delete catalog %s' % catalog_name)

    # write the new catalogs
    for key in catalogs:
        catalogpath = os.path.join("catalogs", key)
        if catalogs[key] != "":
            catalog_data = writePlistToString(catalogs[key])
            try:
                repo.put(catalogpath, catalog_data)
                if output_fn:
                    output_fn("Created %s..." % catalogpath)
            except munkirepo.RepoError as err:
                errors.append(
                    u'Failed to create catalog %s: %s' % (key, err))
        else:
            errors.append(
                "WARNING: Did not create catalog %s because it is empty" % key)

    if icons:
        icon_hashes_plist = os.path.join("icons", "_icon_hashes.plist")
        icon_hashes = writePlistToString(icons)
        try:
            repo.put(icon_hashes_plist, icon_hashes)
            print("Created %s..." % (icon_hashes_plist))
        except munkirepo.RepoError as err:
            errors.append(
                u'Failed to create %s: %s' % (icon_hashes_plist, err))

    # Return any errors
    return errors

File: code/cli/munki/shared/admin/makecatalogslib.swift
//
//  makecatalogslib.swift
//  munki
//
//  Created by Greg Neagle on 6/27/24.
//  Copyright 2024-2026 The Munki Project. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//       https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

import Foundation

enum MakeCatalogsError: Error {
    case PkginfoAccessError(description: String)
    case CatalogWriteError(description: String)
}

struct MakeCatalogOptions {
    var skipPkgCheck: Bool = false
    var force: Bool = false
    var verbose: Bool = false
}

/// Struct that handles building catalogs
struct CatalogsMaker {
    var repo: Repo
    var options: MakeCatalogOptions
    var pkgsinfoList: [String]
    var pkgsList: [String]
    var catalogs: [String: [PlistDict]]
    var errors: [String]
    var warnings: [String]

    init(repo: Repo,
         options: MakeCatalogOptions = MakeCatalogOptions()) async throws
    {
        self.repo = repo
        self.options = options
        catalogs = [String: [PlistDict]]()
        errors = [String]()
        warnings = [String]()
        pkgsinfoList = [String]()
        pkgsList = [String]()
        try await getPkgsinfoList()
        try await getPkgsList()
    }

    /// Returns a list of pkginfo identifiers
    mutating func getPkgsinfoList() async throws {
        do {
            pkgsinfoList = try await repo.list("pkgsinfo")
        } catch is MunkiError {
            throw MakeCatalogsError.PkginfoAccessError(
                description: "Error getting list of pkgsinfo items")
        }
    }

    /// Returns a list of pkg identifiers
    mutating func getPkgsList() async throws {
        do {
            pkgsList = try await repo.list("pkgs")
        } catch is MunkiError {
            throw MakeCatalogsError.PkginfoAccessError(
                description: "Error getting list of pkgs items")
        }
    }

    /// Builds a dictionary containing hashes for all our repo icons
    mutating func hashIcons() async -> [String: String] {
        if options.verbose {
            print("Getting list of icons...")
        }
        var iconHashes = [String: String]()
        if var iconList = try? await repo.list("icons") {
            // remove plist of hashes from the list
            if let index = iconList.firstIndex(of: "_icon_hashes.plist") {
                iconList.remove(at: index)
            }
            for icon in iconList {
                if options.verbose {
                    print("Hashing \(icon)...")
                }
                do {
                    let icondata = try await repo.get("icons/" + icon)
                    iconHashes[icon] = sha256hash(data: icondata)
                } catch let error as MunkiError {
                    errors.append("Error reading icons/\(icon): \(error.description)")
                } catch {
                    errors.append("Unexpected error reading icons/\(icon): \(error)")
                }
            }
        }
        return iconHashes
    }

    /// Returns a case-insensitive match for installer_item from pkgsList, if any
    func caseInsensitivePkgsListContains(_ installer_item: String) -> String? {
        for repo_pkg in pkgsList {
            if installer_item.lowercased() == repo_pkg.lowercased() {
                return repo_pkg
            }
        }
        return nil
    }

    /// Returns true if referenced installer items are present, false otherwise. Updates list of errors.
    mutating func verify(_ identifier: String, _ pkginfo: PlistDict) -> Bool {
        if let installer_type = pkginfo["installer_type"] as? String {
            if ["nopkg", "apple_update_metadata"].contains(installer_type) {
                // no associated installer item (pkg) for these types
                return true
            }
        }
        if !((pkginfo["PackageCompleteURL"] as? String ?? "").isEmpty) {
            // installer item may be on a different server
            return true
        }
        if !((pkginfo["PackageURL"] as? String ?? "").isEmpty) {
            // installer item may be on a different server
            return true
        }

        // Build path to installer item
        let installeritemlocation = pkginfo["installer_item_location"] as? String ?? ""
        if installeritemlocation.isEmpty {
            warnings.append(
                "WARNING: empty or invalid installer_item_location in \(identifier)")
            return false
        }

        // Check if the installer item actually exists
        if !(pkgsList.contains(installeritemlocation)) {
            // didn't find it in the pkgsList; let's look case-insensitive
            if let match = caseInsensitivePkgsListContains(installeritemlocation) {
                warnings.append(
                    "WARNING: \(identifier) refers to installer item: \(installeritemlocation). The pathname of the item in the repo has different case: \(match). This may cause issues depending on the case-sensitivity of the underlying filesystem."
                )
            } else {
                warnings.append(
                    "WARNING: \(identifier) refers to missing installer item: \(installeritemlocation)"
                )
                return false
            }
        }

        // uninstaller checking
        if let uninstalleritemlocation = pkginfo["uninstaller_item_location"] as? String {
            if uninstalleritemlocation.isEmpty {
                warnings.append(
                    "WARNING: empty or invalid uninstaller_item_location in \(identifier)")
                return false
            }
            // Check if the uninstaller item actually exists
            if !(pkgsList.contains(uninstalleritemlocation)) {
                // didn't find it in the pkgsList; let's look case-insensitive
                if let match = caseInsensitivePkgsListContains(uninstalleritemlocation) {
                    warnings.append(
                        "WARNING: \(identifier) refers to uninstaller item: \(uninstalleritemlocation). The pathname of the item in the repo has different case: \(match). This may cause issues depending on the case-sensitivity of the underlying filesystem."
                    )
                } else {
                    warnings.append(
                        "WARNING: \(identifier) refers to missing uninstaller item: \(uninstalleritemlocation)"
                    )
                    return false
                }
            }
        }
        // if we get here we passed all the checks
        return true
    }

    /// Processes pkginfo files and updates catalogs and errors instance variables
    mutating func processPkgsinfo() async {
        catalogs["all"] = [PlistDict]()
        // Walk through the pkginfo files
        for pkginfoName in pkgsinfoList {
            // Try to read the pkginfo file
            var pkginfo = PlistDict()
            let pkginfoIdentifier = "pkgsinfo/" + pkginfoName
            do {
                let data = try await repo.get(pkginfoIdentifier)
                pkginfo = try readPlist(fromData: data) as? PlistDict ?? PlistDict()
            } catch {
                errors.append("Unexpected error reading \(pkginfoIdentifier): \(error)")
                continue
            }
            if !(pkginfo.keys.contains("name")) {
                warnings.append("WARNING: \(pkginfoIdentifier) is missing name key")
                continue
            }
            // don't copy admin notes to catalogs
            if pkginfo.keys.contains("notes") {
                pkginfo["notes"] = nil
            }
            // strip out any keys that start with "_"
            for key in pkginfo.keys {
                if key.hasPrefix("_") {
                    pkginfo[key] = nil
                }
            }
            // sanity checking
            if !options.skipPkgCheck {
                let verified = verify(pkginfoIdentifier, pkginfo)
                if !verified, !options.force {
                    // Skip this pkginfo unless we're running with force flag
                    continue
                }
            }
            // append the pkginfo to the relevant catalogs
            catalogs["all"]?.append(pkginfo)
            if let catalog_list = pkginfo["catalogs"] as? [String] {
                if catalog_list.isEmpty {
                    warnings.append("WARNING: \(pkginfoIdentifier) has an empty catalogs array!")
                } else {
                    for catalog in catalog_list {
                        if !catalogs.keys.contains(catalog) {
                            catalogs[catalog] = [PlistDict]()
                        }
                        catalogs[catalog]?.append(pkginfo)
                        if options.verbose {
                            print("Adding \(pkginfoIdentifier) to \(catalog)...")
                        }
                    }
                }
            } else {
                warnings.append("WARNING: \(pkginfoIdentifier) has no catalogs array!")
            }
        }
        // look for catalog names that differ only in case
        var duplicateCatalogs = [String]()
        for name in catalogs.keys {
            let filtered_lowercase_names = catalogs.keys.filter { $0 != name }.map { $0.lowercased() }
            if filtered_lowercase_names.contains(name.lowercased()) {
                duplicateCatalogs.append(name)
            }
        }
        if !duplicateCatalogs.isEmpty {
            warnings.append(
                "WARNING: There are catalogs with names that differ only by case. " +
                    "This may cause issues depending on the case-sensitivity of the " +
                    "underlying filesystem: \(duplicateCatalogs)")
        }
    }

    /// Clear out old catalogs
    mutating func cleanupCatalogs() async {
        do {
            let catalogList = try await repo.list("catalogs")
            for catalogName in catalogList {
                if !(catalogs.keys.contains(catalogName)) {
                    let catalogIdentifier = "catalogs/" + catalogName
                    do {
                        try await repo.delete(catalogIdentifier)
                    } catch {
                        errors.append("Could not delete catalog \(catalogName): \(error)")
                    }
                }
            }
        } catch {
            errors.append("Could not get list of current catalogs to clean up: \(error)")
        }
    }

    /// Assembles all pkginfo files into catalogs.
    /// User calling this needs to be able to write to the repo/catalogs directory.
    /// Returns a list of any errors it encountered
    mutating func makecatalogs() async {
        // process pkgsinfo items
        await processPkgsinfo()

        // clean up old catalogs no longer needed
        await cleanupCatalogs()

        // write the new catalogs
        for key in catalogs.keys {
            if !(catalogs[key]?.isEmpty ?? true) {
                let catalogIdentifier = "catalogs/" + key
                do {
                    if let value = catalogs[key] {
                        let data = try plistToData(value)
                        try await repo.put(catalogIdentifier, content: data)
                        if options.verbose {
                            print("Created \(catalogIdentifier)...")
                        }
                    }
                } catch let PlistError.writeError(description) {
                    errors.append("Could not serialize catalog \(key): \(description)")
                } catch let error as MunkiError {
                    errors.append("Failed to create catalog \(key): \(error.description)")
                } catch {
                    errors.append("Unexpected error creating catalog \(key): \(error)")
                }
            }
        }

        // make icon hashes
        let iconHashes = await hashIcons()
        // create icon_hashes resource
        if !iconHashes.isEmpty {
            let iconHashesIdentifier = "icons/_icon_hashes.plist"
            do {
                let iconHashesData = try plistToData(iconHashes)
                try await repo.put(iconHashesIdentifier, content: iconHashesData)
                if options.verbose {
                    print("Created \(iconHashesIdentifier)...")
                }
            } catch let PlistError.writeError(description) {
                errors.append("Could not serialize icon hashes: \(description)")
            } catch let error as MunkiError {
                errors.append("Failed to create \(iconHashesIdentifier): \(error.description)")
            } catch {
                errors.append("Unexpected error creating \(iconHashesIdentifier): \(error)")
            }
        }
        return
    }
}
